pkg/controller/corecerts/corecerts_controller.go

```diff
@@ -0,0 +1,228 @@
+	}
+
+	glog.Info("Starting workers")
+	// Launch two workers to process EtcdStorage resources
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	glog.Info("Started workers")
+	<-stopCh
+	glog.Info("Shutting down workers")
+
+	return nil
+}
+
+// runWorker is a long-running function that will continually call the
+// processNextWorkItem function in order to read and process a message on the
+// workqueue.
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem will read a single work item off the workqueue and
+// attempt to process it, by calling the syncHandler.
+func (c *Controller) processNextWorkItem() bool {
+	obj, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	// We wrap this block in a func so we can defer c.workqueue.Done.
+	err := func(obj interface{}) error {
+		// We call Done here so the workqueue knows we have finished
+		// processing this item. We also must remember to call Forget if we
+		// do not want this work item being re-queued. For example, we do
+		// not call Forget if a transient error occurs, instead the item is
+		// put back on the workqueue and attempted again after a back-off
+		// period.
+		defer c.workqueue.Done(obj)
+		var key string
+		var ok bool
+		// We expect strings to come off the workqueue. These are of the
+		// form namespace/name. We do this as the delayed nature of the
+		// workqueue means the items in the informer cache may actually be
+		// more up to date that when the item was initially put onto the
+		// workqueue.
+		if key, ok = obj.(string); !ok {
+			// As the item in the workqueue is actually invalid, we call
+			// Forget here else we'd go into a loop of attempting to
+			// process a work item that is invalid.
+			c.workqueue.Forget(obj)
+			runtime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		// Run the syncHandler, passing it the namespace/name string of the
+		// EtcdStorage resource to be synced.
+		if err := c.syncHandler(key); err != nil {
+			return fmt.Errorf("error syncing '%s': %s", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		c.workqueue.Forget(obj)
+		glog.Infof("Successfully synced '%s'", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		runtime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the EtcdStorage resource
+// with the current status of the resource.
+func (c *Controller) syncHandler(key string) error {
+	secretName := "etcd-core-certs"
+	_, err := c.secretsLister.Secrets(epcNamespace).Get(secretName)
+	if errors.IsNotFound(err) {
+		_, err = c.kubeclientset.CoreV1().Secrets(epcNamespace).Create(c.newSecret(secretName))
```

Comment:
I *think* that this is attempting to create a secret inside of the etcdproxycontroller namespace that contains core etcd connection information.

If that is the case, then I don't think that you should be doing that from within this controller.  Instead, you should assume the existence of the configmap containing the core etcd serving CA bundle and  secret containing the core etcd client cert/key pair for the etcd proxies.  You can then reference them directly by name to mount them into your etcd proxy prod.

In order for you to get the information here in the controller, they had to exist in the namespace as configmaps and secrets before and then get mounted anyway, right?

Reply:
> I think that this is attempting to create a secret inside of the etcdproxycontroller namespace that contains core etcd connection information.

Yes, you're right. This creates a secret containing CA Cert, Client Cert and Client Key.

> If that is the case, then I don't think that you should be doing that from within this controller. Instead, you should assume the existence of the configmap containing the core etcd serving CA bundle and secret containing the core etcd client cert/key pair for the etcd proxies.

Just to confirm I understand you correctly: I shouldn't use this controller to do this, rather, I should assume I have that secret already created in the EtcdProxyController namespace? In that case, I can delete this controller, right?

> In order for you to get the information here in the controller, they had to exist in the namespace as configmaps and secrets before and then get mounted anyway, right?

Yes, they do. Do we want to have some standardized name for now (e.g. `etcd-coreserving-ca` and `etcd-coreclient-cert`), and then later decide about actual names or how are we going to handle those?

Also, @sttts and I talked about this earlier, and we wanted to check can we go only with a Secret, which would contain CA cert, Client Cert and Key, instead of ConfigMap for CA Cert and Secret for Client cert and key. Not sure about drawbacks, but that would be easier to handle and mount, and we would have to watch only one object. What do you think?